src/vm/callsiteinspect.cpp

```diff
@@ -0,0 +1,517 @@
+            _ASSERTE(gc.CurrArgType != NULL);
+            gc.ArgsTypes->SetAt(index, gc.CurrArgType);
+
+            // Record if the argument is ByRef
+            *((UCHAR*)gc.ArgsIsByRef->GetDataPtr() + index) = (!!details.IsByRef);
+        }
+    }
+    GCPROTECT_END();
+
+    // Return details
+    *args = gc.Args;
+    *argsTypes = gc.ArgsTypes;
+    *argsIsByRef = gc.ArgsIsByRef;
+}
+
+void CallsiteInspect::PropagateOutParametersBackToCallsite(
+    _In_ PTRARRAYREF outArgs,
+    _In_ OBJECTREF retVal,
+    _In_ CallsiteDetails &callsite)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_COOPERATIVE;
+    }
+    CONTRACTL_END;
+
+    struct _gc
+    {
+        OBJECTREF RetVal;
+        PTRARRAYREF OutArgs;
+        OBJECTREF CurrArg;
+    } gc;
+    ZeroMemory(&gc, sizeof(gc));
+    gc.OutArgs = outArgs;
+    gc.RetVal = retVal;
+    GCPROTECT_BEGIN(gc);
+    {
+        FramedMethodFrame *frame = callsite.Frame;
+        const INT32 flags = callsite.Flags;
+        MetaSig *pSig = &callsite.MetaSig;
+        pSig->Reset(); // Ensure the sig is in a known state
+
+        // Construct an ArgIterator from the sig
+        ArgIterator argit{ pSig };
+
+        // Propagate the return value only if the call is not a constructor call
+        // and the return type is non-void
+        if ((flags & CallsiteDetails::Ctor) == 0 
+            && pSig->GetReturnType() != ELEMENT_TYPE_VOID)
+        {
+            if (argit.HasRetBuffArg())
+            {
+                // Copy from RetVal into the retBuff.
+                INT64 retVal =  CopyOBJECTREFToStack(
+                                    gc.RetVal,
+                                    *(void**)(frame->GetTransitionBlock() + argit.GetRetBuffArgOffset()),
+                                    pSig->GetReturnType(),
+                                    TypeHandle(),
+                                    pSig,
+                                    TRUE /* should copy */);
+
+                // Copy the return value
+                *(ARG_SLOT *)(frame->GetReturnValuePtr()) = retVal;
+            }
+#ifdef ENREGISTERED_RETURNTYPE_MAXSIZE
+            else if (argit.HasNonStandardByvalReturn())
+            {
+                //
+                // in these cases, we put the pointer to the return buffer into the frame's
+                // return value slot
+                //
+                CopyOBJECTREFToStack(
+                                    gc.RetVal,
+                                    frame->GetReturnValuePtr(),
+                                    pSig->GetReturnType(),
+                                    TypeHandle(),
+                                    pSig,
+                                    TRUE /* should copy */);
+            }
+#endif // ENREGISTERED_RETURNTYPE_MAXSIZE
+            else
+            {
+                // There is no separate return buffer, the retVal should fit in
```

Comment:
Is there any guarantee that this pathway is only useable for IDispatch calls? For other function call types, this isn't true on ARM64/AMD64/X86 in the presence of HFAs and (sometime in the future for ARM64 HVAs)